README.md

```diff
@@ -1,173 +1,276 @@
+
+Update your service definition in `docker-compose.yml` as follows:
+
+```yml
+    your-microservice-name:
+      ...
       ports:
         - 8888:80
-      environment:
-        RACK_ENV: "development"
-      links:
-        - db:database
-      volumes:
-        - /absolute/path/to/your/sources/:/app/
+```
 
+Next, recreate the container by executing
+```bash
+docker-compose up -d your-microservice-name
+```
 
-### Debugging with pry and Better Errors
-Add a breakpoint in your code by inserting a `binding.pry` statement. 
+### Add a dependency to your microservice
+You can install additional dependencies by including a `Gemfile` file next to your `web.rb`. It works as you would expect: just specify the dependencies in the `Gemfile`. They will be installed automatically at build time. In development mode you will need to restart the container.
+
+
+### Execute a SPARQL query
+The template provides several helpers. One of them, `Mu::query`, allows to easily execute a SPARQL query as shown in the following example:
+
+```ruby
+get '/triples' do
+  solutions = Mu::query("SELECT * WHERE { ?s ?p ?o }")
+  triples = solutions.map do |solution|
+    {
+      subject: solution[:s],
+      predicate: solution[:p],
+      object: solution[:o]
+    }
+  end
+  status 200
+  {
+    data: triples
+  }.to_json
+end
+```
 
-When an error occurs, an interactive [Better Errors](https://github.com/charliesome/better_errors) error page is available at `http://{container-ip}/__better_errors`. It's important to access the error page via the container's IP directly and not through localhost, identifier, dispatcher, etc.
+### How to run tests
+To test your app, run the container with `RACK_ENV` set to `test`. All [rspec](http://rspec.info/) tests matching `*_spec.rb` in `spec/` and its subdirectories will be executed.
 
-## Helper methods
-The template provides the user with several helper methods in Sinatra. Some helpers cannot be used outside the Sinatra context.
+    docker run --rm -e RACK_ENV=test microservice-image
 
-#### error(title, status = 400)
-Returns a JSONAPI compliant error response with the given status code (default: `400`).
+To run the tests while developing, start an interactive container in the test enviroment  with your code folder mounted in `/app`:
+
+    docker run --volume /path/to/your/code:/app
+                -e RACK_ENV=test
+                -it semtech/mu-ruby-template:feature-ruby-3 /bin/bash
+
+You can now run your tests inside the container with:
 
-#### graph
+    bundle install
+    rspec
+
+## Reference
+### Framework
+The mu-ruby-template is built on Sinatra. Check [Sinatra's Getting Started guide](https://sinatrarb.com/intro.html) to learn how to build a REST API in Sinatra.
+
+### Utils
+The template offers a `Mu` module with utils to facilitate development
+
+#### Mu::graph
 Returns the application graph configured through the `MU_APPLICATION_GRAPH`.
 
-#### generate_uuid()
+#### Mu::generate_uuid()
```

Comment:
strictly speaking using the `Mu::` module is not required within the web.rb file, since we also register them as helpers. Don't know if that difference needs to be made clear in the readme

Reply:
Added some clarification in the README :+1: